### ISSUE TITLE
[Analyzer] Reverse the order in which Analyzer error message is described

### ISSUE THREAD

**Issue by Martinsos** [wasp-lang/wasp#414]
Right now, Analyzer (compiler) error messages are shown like this:
```
Type error:
  For dictionary field 'dependencies':
    For list element:
      For dictionary field 'version':
        Expected type: string
        Actual type:   number
```

So first we describe the type of the error (`Type Error`), then we describe its context (in dictionary field X, in list, in ...), and then we describe the actual error (`Expected ... Actual ...`).

The thing is however, that this is probably not the optimal way to show error, since we are first starting with less important information and end with the most important information.
We should probably inverse the order, and do smth like:

```
Type error:
  Unexpected type: string
  Actual type:     number
  
    └> For dictionary field 'version'
      └> In list
        └> For dictionary field 'dependencies'
```

This is also how Haskell reports errors: first the error itself, then the context from the inside out.

We should probably check how errors look in couple of other popular compilers/languages, decide on exact format, and then refactor our error message construction to match it.

As for the files where the most of the changes will be happening, those will probably be:
- https://github.com/wasp-lang/wasp/blob/master/waspc/src/Wasp/Analyzer/Parser/ParseError.hs
- https://github.com/wasp-lang/wasp/blob/master/waspc/src/Wasp/Analyzer/TypeChecker/TypeError.hs
- https://github.com/wasp-lang/wasp/blob/master/waspc/src/Wasp/Analyzer/Evaluator/EvaluationError.hs

**Comment by faassen** [comment]
I haven't looked into this in detail yet, but looking into how error reporting in other systems works is quite interesting. Rust is a good candidate to study, and so may be TypeScript, or even Python's static typing system.

**Comment by sodic** [comment]
I've taken a look at several popular languages to see how they approach error reporting.
### Other languages
#### TypeScript
TypeScript takes a top-to-bottom approach when reporting type mismatches.

Let's say we attempt to (incorrectly) use the value `v` in a context that expects a value of type `C`. If `C` is a nested composite type, and the cause of the type error is several levels of nesting deep, TypeScript first reports a mismatch between the actual type of `v` and the expected type `C`. Only then does it "dig down" through type `C` to find the root cause.

The digging/comparing stops as soon as a mismatch is found. A mismatch at the top level is always shown (e.g., imagine assigning string to a nested composite type), while a mismatch at the `N`-th level of nesting is only shown when there's no mismatch at the level `N-1`. In other words, the deeper the error, the longer the message.

```typescript
type Config = {
    dependencies: { lib: string; version: string; }[]
}

const c = {
    dependencies: [
        { lib: 'redux', version: '^4.0.5' },
        { lib: 'react-redux', version: 7 }
    ]
}

const config: Config = c
```
Procudes the following error:
```text
Type '{ dependencies: ({ lib: string; version: string; } | { lib: string; version: number; })[]; }' is not assignable to type 'Config'.
  Types of property 'dependencies' are incompatible.
    Type '({ lib: string; version: string; } | { lib: string; version: number; })[]' is not assignable to type '{ lib: string; version: string; }[]'.
      Type '{ lib: string; version: string; } | { lib: string; version: number; }' is not assignable to type '{ lib: string; version: string; }'.
        Type '{ lib: string; version: number; }' is not assignable to type '{ lib: string; version: string; }'.
          Types of property 'version' are incompatible.
            Type 'number' is not assignable to type 'string'.
```
#### Haskell
As @Martinsos mentioned, Haskell takes a bottom-up approach when reporting type errors. We first see the mismatch/root cause, and then the compiler provides more information to broaden the context.

For example, the code:
```haskell

data Dependency = Dependency {
    name:: [Char],
    version:: [Char]
}


data Config = Config {
    dependencies:: [Dependency]
}

config:: Config
config = Config {
  dependencies = [ 
      Dependency {name = "x", version ="7"},
      Dependency {name = "y", version = 7} 
  ] 
}
```
Produces the error:
```text
errors.hs:16:40: error:
    • No instance for (Num [Char]) arising from the literal ‘7’
    • In the ‘version’ field of a record
      In the expression: Dependency {name = "y", version = 7}
      In the ‘dependencies’ field of a record
   |
16 |       Dependency {name = "y", version =7} 
   |                                        ^
```
#### Go
Go takes the same approach as Haskell. The code:
```go
package main

import "fmt"

type Dependency struct {
	lib     string
	version string
}

type Config struct {
	dependencies []Dependency
}

func main() {
	c := Config{
		dependencies: []Dependency{
			{lib: "asdasd", version: "asdf"},
			{lib: "asdasd", version: 1},
		},
	}
	fmt.Println(c)
}
```
Produces the error:
```
### Verdict
Although the approaches seem different, they

**Comment by sodic** [comment]
**TL;DR:** We should probably reverse the order of messages as this issue suggests. It's consistent with other languages

------

I've taken a look at several popular languages to see how they approach error reporting.
### Other languages
#### TypeScript
TypeScript takes a top-to-bottom approach when reporting type mismatches.

Let's say we attempt to (incorrectly) use the value `v` in a context that expects a value of type `C`. If `C` is a nested composite type, and the cause of the type error is several levels of nesting deep, TypeScript first reports a mismatch between the actual type of `v` and the expected type `C`. Only then does it "dig down" through type `C` to find the root cause.

The digging/comparing stops as soon as a mismatch is found. A mismatch at the top level is always shown (e.g., imagine assigning a string to a nested composite type), while a mismatch at the `N`-th level of nesting is only shown when there's no mismatch at the level `N-1`. In other words, the deeper the error, the longer the message.

<details>
<summary>Show example</summary>

The code
```typescript
type Config = {
    dependencies: { lib: string; version: string; }[]
}

const c = {
    dependencies: [
        { lib: 'redux', version: '^4.0.5' },
        { lib: 'react-redux', version: 7 }
    ]
}

const config: Config = c
```
Procudes the error:
```text
Type '{ dependencies: ({ lib: string; version: string; } | { lib: string; version: number; })[]; }' is not assignable to type 'Config'.
  Types of property 'dependencies' are incompatible.
    Type '({ lib: string; version: string; } | { lib: string; version: number; })[]' is not assignable to type '{ lib: string; version: string; }[]'.
      Type '{ lib: string; version: string; } | { lib: string; version: number; }' is not assignable to type '{ lib: string; version: string; }'.
        Type '{ lib: string; version: number; }' is not assignable to type '{ lib: string; version: string; }'.
          Types of property 'version' are incompatible.
            Type 'number' is not assignable to type 'string'.
```
</details>

#### Haskell
As @Martinsos mentioned, Haskell takes a bottom-up approach when reporting type errors. We first see the mismatch/root cause, and then the compiler provides more information to broaden the context.

<details>

<summary>Show example</summary>

The code:
```haskell

data Dependency = Dependency {
    name:: [Char],
    version:: [Char]
}


data Config = Config {
    dependencies:: [Dependency]
}

config:: Config
config = Config {
  dependencies = [
      Dependency {name = "x", version ="7"},
      Dependency {name = "y", version = 7}
  ]
}
```
Produces the error:
```text
errors.hs:16:40: error:
    • No instance for (Num [Char]) arising from the literal ‘7’
    • In the ‘version’ field of a record
      In the expression: Dependency {name = "y", version = 7}
      In the ‘dependencies’ field of a record
   |
16 |       Dependency {name = "y", version =7}
   |                                        ^
```
</details>

#### Go
Go takes the same approach as Haskell. The code:
<details>

<summary>Show example</summary>

```go
package main

import "fmt"

type Dependency struct {
	lib     string
	version string
}

type Config struct {
	dependencies []Dependency
}

func main() {
	c := Config{
		dependencies: []Dependency{
			{lib: "asdasd", version: "asdf"},
			{lib: "asdasd", version: 1},
		},
	}
	fmt.Println(c)
}
```
Produces the error:
```
./main.go:18:20: cannot use 1 (type untyped int) as type string in field value

Go build failed.
```
</details>

### Verdict
Although the approaches seem different, they are really quite similar.

All languages start by showing where the actual error occurs. However, TypeScript allows us to define nested composite objects without explicitly naming their type, while Go and Haskell don't.

In other words, Go and Haskell _know_ that the true error is at the deepest level because the programmer said so by requiring the nested object to be of type `Dependency`. In other words, Haskell and Go can say: "Hey, this should be a Dependency, but the field type for `version` doesn't match".

TypeScript, on the other hand, doesn't require the programmer to annotate nested objects - it will happily infer the type by looking at its structure. Therefore, the compiler _can't know_ that the true error is at the deepest level. It's perfectly valid to construct the object from the example. What isn't valid, however, is assigning it to a variable of type `Config`. This is what the compiler sees as the true error and where it starts its report.

We can confirm this by slightly changing the TypeScript code above to semantically match the code in Go in Haskell:
```typescript
type Config = {
    dependencies: {
        lib: string;
        version: string;
    }[]
}

// The compiler will now know that the object literal on the right side must
// fit into the type Config
const c: Config = {
    dependencies: [
        {
            lib: 'redux',
            version: '^4.0.5'
        },
        {
            lib: 'react-redux',
            version: 7
        }
    ]
```
The reported error now changes to something similar to what we see in Go and Haskell:
```text
main.ts:16:13 - error TS2322: Type 'number' is not assignable to type 'string'.

16             version: 7
               ~~~~~~~

  main.ts:4:9
    4         version: string;
              ~~~~~~~
    The expected type comes from property 'version' which is declared here on type '{ lib: string; version: string; }'

```

Since Wasp knows that each dependency declaration must have a certain structure, I also think it makes more sense to start by reporting the mismatch at the lowest level. The only difference is that a Wasp user doesn't explicitly say that the dependencies declaration must have this structure, it's baked into the language.

**Comment by Martinsos** [comment]
Awesome analysis @sodic !

Sounds good to me. Do you have any suggestions on the details of the message, what the representation could look like? It would be great if we can make it as readable as possible, so that it can be quickly scanned for the crux of the problem, but that detailed reading also provides more context.

I provided some suggestions at the starting comment in this issue, but I didn't put much research into it, so I am interested to hear what you think might be the best way to do it.

**Comment by sodic** [comment]
I think your suggestion should work well (less is more when it comes to messages). I can't think of anything else useful to say.

**Comment by Martinsos** [comment]
> I think your suggestion should work well (less is more when it comes to messages). I can't think of anything else useful to say.

Ah ok! I thought you might have some ideas how to present it visually, you know, how to show those arrows, how many newlines and where, fun stuff like that :D. Unleash your inner designer :D! Although I do agree, probably best to keep it simple!
